src/plugin.ts

```diff
@@ -0,0 +1,212 @@
+      this.nodeHighlighter.enable();
+    }
+
+    this.iFrameCommunicator.addMessageListener(
+      IFrameMessageType.HighlightsStatus,
+      (data: IHighlightsStatusMessageData) => {
+        if (!data || !this.nodeHighlighter) return;
+
+        if (data.enabled) {
+          this.nodeHighlighter.enable();
+        } else {
+          this.nodeHighlighter.disable();
+        }
+
+        storage.set({
+          highlighterEnabled: data.enabled,
+        });
+      }
+    );
+
+    this.iFrameCommunicator.sendMessage(IFrameMessageType.Initialized, {
+      projectId: this.configuration.projectId,
+      languageCodename: this.configuration.languageCodename,
+      highlighterEnabled,
+    });
+  };
+
+  private watchQueryParams = (): void => {
+    if (isInsideIFrame()) return;
+
+    if (this.state.queryTimerId) {
+      clearTimeout(this.state.queryTimerId);
+    }
+
+    const currentPageUrl = new URL(document.URL);
+    const isPluginEnabled = currentPageUrl.search.includes(this.configuration.queryParam);
+    const hasChanged = this.nodeHighlighter.enabled !== isPluginEnabled;
+
+    if (hasChanged) {
+      if (isPluginEnabled) {
+        this.nodeHighlighter.enable();
+      } else {
+        this.nodeHighlighter.disable();
+      }
+    }
+
+    this.updateState({
+      queryTimerId: setTimeout(this.watchQueryParams, 1000),
+    });
+  };
+
+  private unwatchQueryParams = (): void => {
+    clearTimeout(this.state.queryTimerId);
+  };
+
+  private listenToHighlighterEvents = (): void => {
+    this.nodeHighlighter.addEventListener(NodeHighlighterEventType.ElementClicked, this.onElementClick);
+  };
+
+  private onElementClick = (elementClickData: Partial<IElementClickMessageData>): void => {
+    const data: Partial<IElementClickMessageData> = {
+      ...elementClickData,
+      projectId: elementClickData.projectId || this.configuration.projectId || undefined,
+      languageCodename: elementClickData.languageCodename || this.configuration.languageCodename || undefined,
+    };
+
+    let errorsCount = 0;
+
+    if (!data.projectId) {
+      console.error('Warning: project ID is required to handle element click.');
+      errorsCount++;
+    }
+
+    if (!data.languageCodename) {
+      console.error('Warning: language codename is required to handle element click.');
+      errorsCount++;
+    }
+
+    if (!data.itemId) {
+      console.error('Warning: item ID is required to handle element click.');
+      errorsCount++;
+    }
+
+    if (!data.elementCodename) {
+      console.error('Warning: element codename is required to handle element click.');
+      errorsCount++;
+    }
+
+    if (errorsCount > 0) return;
+
+    if (isInsideIFrame() && this.iFrameCommunicator) {
+      this.iFrameCommunicator.sendMessage(IFrameMessageType.ElementClicked, data as IElementClickMessageData);
+    } else {
+      const link = buildElementLink(data.projectId!, data.languageCodename!, data.itemId!, data.elementCodename!);
```

Comment:
These `!` are nasty, you should probably make some type validator function `isCompleteElementClickMessageData` instead of this nasty casting ...